mlflow/langchain/_langchain_autolog.py

```diff
@@ -180,6 +185,439 @@ def _chain_with_retriever(model):
+            parent_run_id: Optional[UUID] = None,
+            metadata: Optional[Dict[str, Any]] = None,
+            run_type: Optional[str] = None,
+            name: Optional[str] = None,
+            **kwargs: Any,
+        ):
+            """Start span for a chain run."""
+            if metadata:
+                kwargs.update({"metadata": metadata})
+            # not considering streaming events for now
+            self._start_span(
+                span_name=name or "chain",
+                parent_run_id=parent_run_id,
+                span_type=SpanType.CHAIN,
+                run_id=run_id,
+                inputs=self._get_chain_inputs(inputs),
+                attributes=kwargs,
+            )
+
+        @override
+        def on_chain_end(
+            self,
+            outputs: Dict[str, Any],
+            *,
+            run_id: UUID,
+            inputs: Optional[Union[Dict[str, Any], Any]] = None,
+            **kwargs: Any,
+        ):
+            """Run when chain ends running."""
+            chain_span = self._get_span_by_run_id(run_id)
+            if inputs:
+                chain_span.set_inputs(self._get_chain_inputs(inputs))
+            self._end_span(chain_span, outputs=outputs)
+
+        @override
+        def on_chain_error(
+            self,
+            error: BaseException,
+            *,
+            inputs: Optional[Union[Dict[str, Any], Any]] = None,
+            run_id: UUID,
+            **kwargs: Any,
+        ):
+            """Run when chain errors."""
+            chain_span = self._get_span_by_run_id(run_id)
+            if inputs:
+                chain_span.set_inputs(self._get_chain_inputs(inputs))
+            chain_span.add_event(ExceptionEvent(error))
+            self._end_span(chain_span, status=SpanStatus(TraceStatus.ERROR, str(error)))
+
+        @override
+        def on_tool_start(
+            self,
+            serialized: Dict[str, Any],
+            input_str: str,
+            *,
+            run_id: UUID,
+            tags: Optional[List[str]] = None,
+            parent_run_id: Optional[UUID] = None,
+            metadata: Optional[Dict[str, Any]] = None,
+            name: Optional[str] = None,
+            inputs: Optional[Dict[str, Any]] = None,
+            **kwargs: Any,
+        ):
+            """Start span for a tool run."""
+            if metadata:
+                kwargs.update({"metadata": metadata})
+            self._start_span(
+                span_name=name or "tool",
+                parent_run_id=parent_run_id,
+                span_type=SpanType.TOOL,
+                run_id=run_id,
+                inputs={"input_str": input_str},
+                attributes=kwargs,
+            )
+
+        @override
+        def on_tool_end(self, output: Any, *, run_id: UUID, **kwargs: Any):
+            """Run when tool ends running."""
+            tool_span = self._get_span_by_run_id(run_id)
+            self._end_span(tool_span, outputs={"output": str(output)})
```

Comment:
nit: now span `outputs` support arbitral object not only dict, so we don't need to wrap single input/output with a dict.